dm/syncer/syncer.go

```diff
@@ -1882,12 +1882,17 @@ func (s *Syncer) Run(ctx context.Context) (err error) {
 					continue
 				}
 			}
-			// set endLocation.Suffix=0 of last replace event
+			// set endLocation.Suffix=0 of last replace or inject event
 			// also redirect stream to next event
 			if currentLocation.Suffix > 0 && e.Header.EventSize > 0 {
 				currentLocation.Suffix = 0
-				s.isReplacingErr = false
-				err = s.streamerController.RedirectStreamer(tctx, currentLocation)
+				s.isReplacingOrInjectingErr = false
+				if s.errOperatorHolder.IsInject(startLocation) {
+					s.errOperatorHolder.SetHasAllInjected(startLocation)
+					err = s.streamerController.RedirectStreamer(tctx, startLocation)
```

Comment:
seem in current impl inject == repleace ? or could you point out where is the inject event to binlog stream logic?

Reply:
replace will jump a event, but inject will not. so that replace redirect to `currentLocation` , inject redirect to `startLocation`.

Reply:
i still can't understand, could you please explain more?

i see in L1871 `currentLocation = startLocation` so semms both inject and replace will redirect to `startLocation` ?

i thinks `inject` means **inject** one event between two old events, eg:
```
raw:
event1 -> event2

after injected:
event1 -> injected event -> event2
```

Reply:
ok, `inject` command like `binlog [-b binlog-pos] inject <task-name> <ddl>`, it will inject `ddl` into some ` binlog-pos`. so the most important logic is when we sync at this ` binlog-pos`, we will sync ddl which is in `inject` command, after sync that ddl or ddls,  we will execute  original `ddl` at the `binlog-pos` and this is diffrent with `replace`.

Reply:
> 

yes, i will add integration test later.

Reply:
thanks so much, now i can understand!